File: sdk/python/symbolchain/core/Bip32.py
import hashlib
import hmac

from mnemonic import Mnemonic

from .BufferWriter import BufferWriter
from .CryptoTypes import PrivateKey


class Bip32Node():
    """Representation of a BIP32 node."""

    def __init__(self, key, data):
        """Creates a BIP32 node around a key and data."""
        hmac_result = hmac.new(key, data, hashlib.sha512).digest()

        self.private_key = PrivateKey(hmac_result[0:PrivateKey.SIZE])
        self.chain_code = hmac_result[PrivateKey.SIZE:]

    def derive_one(self, identifier):
        """Derives a direct child node with specified identifier."""
        hmac_data_writer = BufferWriter('big')
        hmac_data_writer.write_int(0, 1)
        hmac_data_writer.write_bytes(self.private_key.bytes)
        hmac_data_writer.write_int(0x80000000 | identifier, 4)
        return Bip32Node(self.chain_code, hmac_data_writer.buffer)

    def derive_path(self, path):
        """Derives a descendent node with specified path."""
        next_node = self
        for identifier in path:
            next_node = next_node.derive_one(identifier)

        return next_node


class Bip32():
    """Factory of BIP32 root nodes """

    def __init__(self, curve_name='ed25519', mnemonic_language='english'):
        """Creates a BIP32 root node factory."""
        self.root_hmac_key = (curve_name + ' seed').encode('utf8')
        self.mnemonic_language = mnemonic_language

    def from_seed(self, seed):
        """Creates a BIP32 root node from a seed."""
        return Bip32Node(self.root_hmac_key, seed)

    def from_mnemonic(self, mnemonic, password):
        """Creates a BIP32 root node from a BIP39 mnemonic and password."""
        return self.from_seed(Mnemonic(self.mnemonic_language).to_seed(mnemonic, password))

File: sdk/javascript/src/Bip32.js
import { PrivateKey } from './CryptoTypes.js';
import Mnemonic from 'bitcore-mnemonic';
import crypto from 'crypto';

/**
 * Representation of a BIP32 node.
 */
export class Bip32Node {
	/**
	 * Creates a BIP32 node around a key and data.
	 * @param {Uint8Array} hmacKey BIP32 HMAC key.
	 * @param {Uint8Array} data BIP32 seed.
	 */
	constructor(hmacKey, data) {
		const hmac = crypto.createHmac('sha512', hmacKey);
		hmac.update(data);
		const hmacResult = hmac.digest();

		/**
		 * Private key associated with this node.
		 * @type PrivateKey
		 */
		this.privateKey = new PrivateKey(hmacResult.subarray(0, PrivateKey.SIZE));

		/**
		 * Chain code associated with this node.
		 * @type Uint8Array
		 */
		this.chainCode = hmacResult.subarray(PrivateKey.SIZE);
	}

	/**
	 * Derives a direct child node with specified identifier.
	 * @param {number} identifier Child identifier.
	 * @returns {Bip32Node} BIP32 child node.
	 */
	deriveOne(identifier) {
		const childData = new Uint8Array(1 + PrivateKey.SIZE + 4);
		childData[0] = 0;
		childData[childData.length - 4] = 0x80;

		for (let i = 0; 4 > i; ++i)
			childData[childData.length - 1 - i] |= (identifier >> (8 * i)) & 0xFF;

		for (let i = 0; i < PrivateKey.SIZE; ++i)
			childData[1 + i] = this.privateKey.bytes[i];

		return new Bip32Node(this.chainCode, childData);
	}

	/**
	 * Derives a descendent node with specified path.
	 * @param {Array<number>} path BIP32 path.
	 * @returns {Bip32Node} BIP32 node at the end of the path.
	 */
	derivePath(path) {
		/** @type Bip32Node */
		let nextNode = this;
		path.forEach(identifier => {
			nextNode = nextNode.deriveOne(identifier);
		});

		return nextNode;
	}
}

/**
 * Factory of BIP32 root nodes.
 */
export class Bip32 {
	/**
	 * Creates a BIP32 root node factory.
	 * @param {string} curveName Elliptic curve to use.
	 * @param {string} mnemonicLanguage Language of constructed mnemonics.
	 */
	constructor(curveName = 'ed25519', mnemonicLanguage = 'english') {
		/**
		 * @private
		 */
		this._rootHmacKey = Buffer.from(`${curveName} seed`);

		/**
		 * @private
		 */
		this._mnemonicLanguage = mnemonicLanguage;
	}

	/**
	 * Creates a BIP32 root node from a seed.
	 * @param {Uint8Array} seed BIP32 seed.
	 * @returns {Bip32Node} BIP32 root node.
	 */
	fromSeed(seed) {
		return new Bip32Node(this._rootHmacKey, seed);
	}

	/**
	 * Creates a BIP32 root node from a BIP39 mnemonic and password.
	 * @param {string} mnemonic BIP32 mnemonic.
	 * @param {string} password BIP32 mnemonic password.
	 * @returns {Bip32Node} BIP32 root node.
	 */
	fromMnemonic(mnemonic, password) {
		const wordlist = Mnemonic.Words[this._mnemonicLanguage.toUpperCase()];
		return this.fromSeed(new Mnemonic(mnemonic, wordlist).toSeed(password));
	}

	/**
	 * Creates a random BIP32 mnemonic.
	 * @param {number} seedLength Length of random seed to use when generating mnemonic.
	 * @returns {string} Random mnemonic created with the specified entropy.
	 */
	random(seedLength = 32) {
		const wordlist = Mnemonic.Words[this._mnemonicLanguage.toUpperCase()];
		return new Mnemonic(seedLength * 8, wordlist).phrase;
	}
}
